### ISSUE TITLE
Signout with ReturnURL doesn't work as expected ? 

### ISSUE THREAD

**Issue by ExtremeSwat** [DuendeSoftware/Support#1183]
**Which version of Duende BFF are you using?**
2.1.1
**Which version of .NET are you using?**
7.0
**Describe the bug**
When logging out, everything works fine except I'm not getting redirected.

There is no redirection happening

**To Reproduce**

Doing something like this from my front-end
```
public logout(): void {
    this.getLogoutUrl().subscribe(
      (urllogout) => {
        urllogout += `&returnUrl=${window.location.origin}`;

        if (urllogout) {
          window.location.replace(urllogout);
        }
      }
    );
  }
```

I can see in my BFF's logs that the return url is being 'used'

**Expected behavior**
Expecting for the redirection to occur post-logout, to my front-end.

**Log output/exception with stacktrace**
```
[13:38:47 Debug] Duende.Bff.DefaultLogoutService
Processing logout request

[13:38:47 Debug] Duende.Bff.ServerSideTicketStore
Removing AuthenticationTicket from store for key 2C1DCE8E2E424480C652BDD0891D4CAAAF66106C4D5F6A3EBF9004615B3624FC

[13:38:47 Debug] Duende.Bff.EntityFramework.UserSessionStore
Deleting user session record in store for sub 7becac79-c29d-4093-897c-e5671dbad0df sid 26FA8167C07352B19539D4AF9FD475F9

[13:38:47 Debug] Duende.Bff.PostConfigureApplicationCookieRevokeRefreshToken
Revoking user's refresh tokens in OnSigningOut for subject id: 7becac79-c29d-4093-897c-e5671dbad0df

[13:38:47 Information] Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationHandler
AuthenticationScheme: cookie signed out.

[13:38:47 Debug] Duende.Bff.DefaultLogoutService
Logout endpoint triggering SignOut with returnUrl https://staging.adcdtsdev.com

[13:38:47 Information] Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler
AuthenticationScheme: oidc signed out.
2024-03-26T13:38:47.790531623Z
```

So it does seem to be working fine, the redirection doesn't occur, I just 'arrive' on my STS page where I am greeted by 'You've logged out' (using Skoruba's implementation)

Maybe do I have to catch in MVC that returnURL and do the return from there maybe ? 🤔

```
data
```

**Additional context**

**Comment by RolandGuijt** [comment]
Do you have a PostLogoutUri configured in the client configuration of your identity provider? 
If you are using IdentityServer the interaction service [can be used](https://github.com/DuendeSoftware/Samples/blob/main/IdentityServer/v7/Quickstarts/2_InteractiveAspNetCore/src/IdentityServer/Pages/Account/Logout/LoggedOut.cshtml.cs) to get that Uri and to [show a link](https://github.com/DuendeSoftware/Samples/blob/main/IdentityServer/v7/Quickstarts/2_InteractiveAspNetCore/src/IdentityServer/Pages/Account/Logout/LoggedOut.cshtml) to redirect back.
To get redirected automatically in our quickstarts a little javascript file is loaded to take care of the redirect.
In the quickstart it is enabled [here](https://github.com/DuendeSoftware/Samples/blob/main/IdentityServer/v6/Quickstarts/2_InteractiveAspNetCore/src/IdentityServer/Pages/Account/Logout/LogoutOptions.cs) and loaded [here](https://github.com/DuendeSoftware/Samples/blob/main/IdentityServer/v6/Quickstarts/2_InteractiveAspNetCore/src/IdentityServer/Pages/Account/Logout/LoggedOut.cshtml).

Please let me know if you have something similar to this. If so we can of course investigate further.

**Comment by AndersAbel** [comment]
@ExtremeSwat Do you have any updates on this? If the problem is solved we'd like to close the issue.

**Comment by ExtremeSwat** [comment]
Apologies for the huge delay, will do some testing and provide some feedback asap.

**Comment by ExtremeSwat** [comment]
Hey, again, apologies.

Yeah, we do have the same things, so according to ```If you are using IdentityServer the interaction service [can be used](https://github.com/DuendeSoftware/Samples/blob/main/IdentityServer/v7/Quickstarts/2_InteractiveAspNetCore/src/IdentityServer/Pages/Account/Logout/LoggedOut.cshtml.cs) to get that Uri and to [show a link](https://github.com/DuendeSoftware/Samples/blob/main/IdentityServer/v7/Quickstarts/2_InteractiveAspNetCore/src/IdentityServer/Pages/Account/Logout/LoggedOut.cshtml) to redirect back.``` we have this:

```
    private async Task<LoggedOutViewModel> BuildLoggedOutViewModelAsync(string logoutId)
    {
        // get context information (client name, post logout redirect URI and iframe for federated signout)
        var logout = await _interaction.GetLogoutContextAsync(logoutId);

        await _identityAuditService.SendAuditEventAsync(logout?.SubjectId, logout?.ClientName, logout?.ClientId, "User logout", "STS.Identity", 
            new { PostLogoutRedirectUri = logout?.PostLogoutRedirectUri, SignoutIFrameUrl = logout?.SignOutIFrameUrl });

        _logger.LogInformation($"[BuildLoggedOutViewModelAsync] - Subject id is: {logout?.SubjectId}.");
        _logger.LogInformation($"[BuildLoggedOutViewModelAsync] - Client name is: {logout?.ClientName}.");
        _logger.LogInformation($"[BuildLoggedOutViewModelAsync] - Client id is: {logout?.ClientId}.");
        _logger.LogInformation($"[BuildLoggedOutViewModelAsync] - PostLogoutRedirectUri is: {logout?.PostLogoutRedirectUri}.");
        _logger.LogInformation($"[BuildLoggedOutViewModelAsync] - SignOutIFrameUrl is: {logout?.SignOutIFrameUrl}.");
        if (logout?.Parameters.Count > 0)
        {
            int count = 0;
            foreach (var param in logout?.Parameters)
            {
                count++;
                _logger.LogInformation($"[BuildLoggedOutViewModelAsync] - Parameter {count} is: {param}.");
            }
            
        }

        var vm = new LoggedOutViewModel
        {
            AutomaticRedirectAfterSignOut = AccountOptions.AutomaticRedirectAfterSignOut,
            PostLogoutRedirectUri = string.IsNullOrEmpty(logout?.PostLogoutRedirectUri) ? _logoutConfiguration?.RedirectUrl : logout?.PostLogoutRedirectUri,
            ClientName = string.IsNullOrEmpty(logout?.ClientName) ? logout?.ClientId : logout?.ClientName,
            SignOutIframeUrl = logout?.SignOutIFrameUrl,
            LogoutId = logoutId
        };

        if (User?.Identity.IsAuthenticated == true)
        {
            var idp = User.FindFirst(JwtClaimTypes.IdentityProvider)?.Value;
            if (idp != null && idp != IdentityServerConstants.LocalIdentityProvider)
            {
                var providerSupportsSignout = await HttpContext.GetSchemeSupportsSignOutAsync(idp);
                if (providerSupportsSignout)
                {
                    if (vm.LogoutId == null)
                    {
                        // if there's no current logout context, we need to create one
                        // this captures necessary info from the current logged in user
                        // before we signout and redirect away to the external IdP for signout
                        vm.LogoutId = await _interaction.CreateLogoutContextAsync();
                    }

                    vm.ExternalAuthenticationScheme = idp;
                }
            }
        }

        return vm;
    }
}
```

The weird thing is that ```SignOutIFrameUrl``` is null, we've reached that point where we've started logging everything, everywhere since we can't explain what's missing.

As in for the RazorPage for the LoggedOut component:

```
@using Microsoft.AspNetCore.Mvc.Localization
@inject IViewLocalizer Localizer
@model HTC.IdentityServer.STS.Identity.ViewModels.Account.LoggedOutViewModel

@{
	// set this so the layout rendering sees an anonymous user
	ViewData["signed-out"] = true;
}

<div class="page-header logged-out-container">
	<h1>
		@Localizer["Title"]
		<small>@Localizer["SubTitle"]</small>
	</h1>

	@if (Model.PostLogoutRedirectUri != null)
	{
		<div>
			@Localizer["Click"] <a class="PostLogoutRedirectUri" href="@Model.PostLogoutRedirectUri">@Localizer["Here"]</a> @Localizer["Return"]
			<span>@Model.ClientName</span>.
		</div>
	}

	@if (Model.SignOutIframeUrl != null)
	{
		<iframe width="0" height="0" class="signout" src="@Model.SignOutIframeUrl"></iframe>
	}
</div>

@section scripts
	{
	@if (Model.AutomaticRedirectAfterSignOut)
	{
		<script src="~/js/signout-redirect.js"></script>
	}
}
```

From what I can see, based on your samples, our only difference is in the name of the file, it's called ```AccountOptions``` on our side.

```
  public class AccountOptions
  {
      public static bool AllowLocalLogin = true;
      public static bool AllowRememberLogin = true;
      public static TimeSpan RememberMeLoginDuration = TimeSpan.FromDays(30);

      public static bool ShowLogoutPrompt = false;
      public static bool AutomaticRedirectAfterSignOut = true;

      public static string InvalidCredentialsErrorMessage = "Invalid username or password";
      public static string PasswordExpiredMessage = "The password has expired";
  }
  ```

**Comment by RolandGuijt** [comment]
Can you please verify that the javascript file "signout-redirect.js" is actually loaded by the browser? 
A great way to do that is to use the network tab of the F12 developer tools.
I'm asking because it might be missing from your setup or somehow the AutomaticRedirectAfterSignOut boolean could end up being false in the view. A good way to check that would be to set a breakpoint in the Razor of the LoggedOut page.

**Comment by ExtremeSwat** [comment]
Heya, the redirection works, just that there is something wrong with the binding, we use Duende auth for both Staging and Dev environments and basically rather than redirecting me to staging.xxxx (what BFF is asking) it's redirecting to htx.xxx because it's a hardcoded value I've tried out to see if redirection even occurs.

So in my code i have this line
```PostLogoutRedirectUri = string.IsNullOrEmpty(logout?.PostLogoutRedirectUri) ? _logoutConfiguration?.RedirectUrl : logout?.PostLogoutRedirectUri,```

It always goes to the defaulted value because PostLogoutRedirectUri is always 'empty'

**Comment by RolandGuijt** [comment]
I've asked this before but as far as I know you didn't answer the question: what is the value for the PostLogoutRedirectUris collection in the client config on the identity provider side? Please post the complete client config so we have the complete picture.

**Comment by ExtremeSwat** [comment]
Hello @RolandGuijt, again, apologies for answering so late.

We're using a client called ```bff-client```

![image](https://github.com/DuendeSoftware/Support/assets/9104188/098417e7-66ed-403d-8ec0-83d26137eb79)

![image](https://github.com/DuendeSoftware/Support/assets/9104188/6a70e19f-1195-4a2d-b31e-ef9dd5a66756)

TLDR:
* https://localhost:4200/
* https://staging.adcdtsdev.com/

So in my BFF this is my config:
```
// cookie options
builder.Services.AddAuthentication(options =>
    {
        options.DefaultScheme = "cookie";
        options.DefaultChallengeScheme = "oidc";
        options.DefaultSignOutScheme = "oidc";
    })
    .AddCookie("cookie", options =>
    {
        // host prefixed cookie name
        options.Cookie.Name = "__Host-spa-ef";

        options.Cookie.SameSite = SameSiteMode.Strict;
    })
    .AddOpenIdConnect("oidc", options =>
    {
        options.Authority = settings.Authority;

        // confidential client using code flow + PKCE
        options.ClientId = "bff-client";
        options.ClientSecret = "secret";

        options.ResponseType = "code";
        options.ResponseMode = "query";

        options.SignedOutRedirectUri = settings.FrontEndHost;

        options.MapInboundClaims = false;
        options.GetClaimsFromUserInfoEndpoint = true;

        // since we're in AKS we could disable it and mind our own business 

        // test
        if (builder.Environment.IsDevelopment())
            options.RequireHttpsMetadata = false;

        options.SaveTokens = true;

        // request scopes + refresh tokens
        options.Scope.Clear();
        options.Scope.Add("openid");
        options.Scope.Add("profile");
        options.Scope.Add("webapi");
        options.Scope.Add("offline_access");
        options.Scope.Add("MyClientId_api");
    });
    ```
    
    Not sure what I'm missing

**Comment by RolandGuijt** [comment]
You might be confused here about how the logout mechanism works with BFF:
As you know, there are 3 applications in play: the front-end (js app), the BFF (ASP.NET Core) and the identity provider.
Logging out is a multi-stage process:

First the front-end calls the logout endpoint on the BFF that has to include the session id for safety. The url is by default contained in the bff:logout_url claim. By default the BFF will redirect to / when the logout completes, but this can be overridden with the returnUrl query parameter. Look [here](https://docs.duendesoftware.com/identityserver/v7/bff/session/management/logout/) for examples.

When the BFF receives the logout request it will call the [end_session endpoint](https://docs.duendesoftware.com/identityserver/v7/reference/endpoints/end_session/) on the identity provider. This includes the post logout redirect uri that must match with one of the configured post logout redirect uris in the client configuration of the identity provider. If there is no match that's a show stopper.
Important to understand here that this post logout redirect uri is different from the returnUrl passed in by the front-end. This one is calling back to the BFF and has nothing to do with the front-end. 
The OpenIdConnect handler in the BFF will automatically provide an endpoint for this callback. The default for that is https://[domain]/signout-callback-oidc. There is no need to set options.SignedOutRedirectUri. The BFF knows about the URL by reading the OpenIdConnect config.
You can see an example on how this is configured in IdentityServer [here](https://github.com/DuendeSoftware/BFF/blob/main/samples/IdentityServer/Config.cs).

After the callback is done from identity provider to BFF, the BFF will get rid of the identity cookie and call the returnUrl described in the second paragraph.

With this info you should be able to get it working.

**Comment by ExtremeSwat** [comment]
I will give  it a try asap and let you know, thanks !

**Comment by RolandGuijt** [comment]
OK. Closing for now but feel free to continue this if you encounter issues.

**Comment by apetrut** [comment]
@RolandGuijt I eventually managed to fix it. I had to add the Client post logout redirect uri in this format: https://[subdomain].[domain]/signout-callback-oidc.

If I only added the domain it wouldn't work.
